colossalai/zero/gemini/gemini_optimizer.py

```diff
@@ -137,9 +143,8 @@ def _set_grad_ptr(self):
                 begin, end = self.param_to_range[fake_param]
                 chunk16 = chunk32.paired_chunk
 
-                fake_param.data = chunk16.payload[begin:end]
-                fake_param.grad = fake_param.data
                 fake_param.data = chunk32.payload[begin:end]
+                fake_param.grad = chunk16.payload[begin:end].to(chunk32.dtype)
```

Comment:
Why cast grad to fp32 here? This will use extra memory, as hybrid adam will cast grad to fp32 one by one during optimizing.

Reply:
```suggestion
                grad_payload = chunk16.payload[begin:end]
                if self.untested_optim_flag:
                    grad_payload = grad_payload.to(chunk32.dtype)         
                fake_param.grad = grad_payload
```

Introduced casting to work with normal torch optimizers (the purpose of this PR). Normal PyTorch optimizers do not accept FP16 gradients and FP32 parameters. I cast the type to update FP32 via FP16 gradients as the [low level zero optimizer does](https://github.com/hpcaitech/ColossalAI/blob/f83ea813f5bb6ecc5597c7f1bf97870d46de1c49/colossalai/zero/low_level/low_level_optim.py#LL468C21-L468C21). Since Colossal's optimizer supports hybrid tensors, why not just cast gradients for untested optimziers? Please suggest me if there is a better way. thank you.